src/main/java/com/google/sps/data/algorithm/TspSolver.java

```diff
@@ -0,0 +1,414 @@
+   * 
+   * @param center central location
+   * @param pois list of pois to visit
+   */
+  public void solver(String center, List<String> pois) throws IOException {
+    if (pois == null) {
+      throw new IllegalArgumentException("pois input list is null");
+    } else if (pois.isEmpty()) {
+      throw new IllegalArgumentException("pois input list is empty");
+    } else if (center == null) {
+      throw new IllegalArgumentException("center poi is null");
+    } else if (center.isEmpty()) {
+      throw new IllegalArgumentException("center poi is empty string");
+    }
+
+    // variables used for tracking
+    int currPos = START_POS;
+    int numNodes = pois.size() + 1; // total nodes: #locations + center
+    LocalTime currentTime = START_TIME;
+    int count = START_COUNT;
+    Tuple cost = new Tuple(START_COST, new ArrayList<>());
+    Tuple ans = new Tuple(START_ANS, new ArrayList<>());
+    boolean[] visited = new boolean[numNodes]; // sets values to "false"  
+
+    // call function to set timeMatrix, intToPlaceId, & openHours
+    createTimeMatrix(center, pois);
+    populateIntMapAndOpenHours(center, pois);
+
+    // set central visited node to visted
+    visited[0] = true;
+
+    // call recursive helper fuction to get final Answer
+    this.finalAnswer = solverHelper(currPos, numNodes, currentTime, count, 
+        cost, ans, visited);
+  }
+
+  /**
+   * Recursive helper function that is used to solve computational part of
+   * algorithm. Called in global `solver()` method
+   *
+   * @param currPos represents the placeId of where the alg currently is
+   * @param numNodes total number of locations to visit (includes center)
+   * @param currentTime time if path at this point were taken 
+   * @param count number of pois visited during this path
+   * @param cost Tuple of current path (time cost, List of pois)
+   * @param ans the current best solution found
+   * @param array of which pois have been visited by this path
+   */
+  public Tuple solverHelper(int currPos, int numNodes, LocalTime currentTime, 
+      int count, Tuple cost, Tuple ans, boolean[] visited) {
+    /**
+     * If last node is reached and shares an edge w/ start node
+     * Calculate min of cost and currAns
+     * return and keep traversing the graph/matrix
+     */
+    if ((count == numNodes-1) && timeMatrix[currPos][0] > 0) {
+      if (ans.getCurrAns() < cost.getCurrAns() + timeMatrix[currPos][0]) {
+        return ans;
+      } else {
+        cost.incCurrAns(timeMatrix[currPos][0]);
+        return cost;
+      }
+    }
+
+    /**
+     * Backtracking Step: loop through the time Matrix from currPos, increase
+     * count by 1 and update (int) and (List) value of cost
+     */
+    for (int i = 0; i < numNodes; i++) {
+      // set open boolean and time to add to currenTime
+      boolean isOpen = (boolean) checkIsOpenSetTime(currentTime, i,
+          currPos).get(0);
```

Comment:
meant to include this in the review but this check is great because it lets you quit early as soon as your route arrives somewhere after closing. so, eg: if the second destination would be closed when you arrive this algorithm doesn't waste any time considering what would happen after that in the route. Nice!